I heard you like inferring types but don't like global state...

jrevels opened the pull request:
...so @vtjnash and I removed the work queue by refactoring inference to run recursively. 

I'm not sure yet how far this will make it through CI. @vtjnash will probably be able to answer any algorithmic questions better than myself - this is my first foray into the dangerous world of type inference. I mainly sat there like a monkey poking at the keyboard.

Besides reducing dependence on global state, this PR provides the following goodies:

- exactly precise cycle detection
- more optimal inference state provided to inlining pass
- Jameson says it might be faster (still need to time this)

This PR enables the following future work:

- a more optimal inlining order
- effect-free type inference
- more optimal/reliable inference recursion detection rules (will allow automatic effect-free computation and constant propagation)

@vtjnash is sad that he has to write a new blog post now.

jrevels committed:
first steps towards recursive type inference

jrevels committed:
fill in mocked implementation of cyclic frame retrieval + intermediate caller merging

jrevels committed:
remove dead code

jrevels committed:
fix add_backedge

jrevels committed:
remove unused global_sv

jrevels committed:
reimplement typeinf in the new recursive world

jrevels committed:
clean up old variable usage

jrevels committed:
add back in age convergence logic

jrevels committed:
fix those nasty nasty bugs

vtjnash committed:
correction to merge_call_chain algo (missing initial edge that makes the loop)

vtjnash committed:
this is necessary

jrevels committed:
callers --> callers_in_cycle

vtjnash committed:
restore heuristic limiting

JeffBezanson commented:
Cool!

Does this do anything to avoid using a very large amount of stack space?

JeffBezanson commented on base/inference.jl:
```diff
@@ -296,10 +295,7 @@ end
-const nactive = Array{Int,0}()
-nactive[] = 0
-const workq = Vector{InferenceState}() # set of InferenceState objects that can make immediate progress
+# None! There is none! :)
```
:-)

JeffBezanson commented on base/inference.jl:
```diff
@@ -1287,26 +1283,13 @@ function abstract_call_gf_by_type(f::ANY, atype::ANY, sv::InferenceState)
-
-        # limit argument type size growth
         if mightlimitlength || mightlimitdepth
             # TODO: FIXME: this heuristic depends on non-local state making type-inference unpredictable
```
Does this change fix this issue?

jrevels replied:
Unfortunately not, but this PR does enable a code rewrite to fix the issue. This is also relevant to your other question:

> Does this do anything to avoid using a very large amount of stack space?

In short, nope. @vtjnash believes that the future rewrite of the call recursion limit code here will help with stack space usage.

JeffBezanson commented on base/inference.jl:
```diff
@@ -2415,42 +2374,76 @@ function code_for_method(method::Method, atypes::ANY, sparams::SimpleVector, wor
+    return frame
+end
+
+function in_egal(item, collection)
```
Already exists in this file; called `contains_is`.

jrevels committed:
in_egal --> contains_is

martinholters commented:
Looks like https://github.com/JuliaLang/julia/blob/970a7427b62cf710db216432518290c24a465465/test/staged.jl#L150 is the only problem hit by CI? That would be pretty awesome!

jrevels committed:
remove test checking the state of the now-nonexistant inference work queue

jrevels commented:
> Looks like https://github.com/JuliaLang/julia/blob/970a7427b62cf710db216432518290c24a465465/test/staged.jl#L150 is the only problem hit by CI? That would be pretty awesome!

That would be crazy! I removed the now-invalid reference to the work queue (and kept the rest of that tiny module), so I guess we'll see what happens. 

Let's see if this can make it through a whole BaseBenchmarks run:

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21677#issuecomment-298929159) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8a533d05df1770d6af1c52d24d404d28aea86fa7/6893f00_vs_df66fe6/report.md). cc @jrevels

martinholters commented:
This seems to go so smooth, it's almost scary.

jrevels commented:
Could anybody hazard a guess as to what's causing the [`Base.Test.TestSetException` on 32-bit Windows](https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.13610/job/iex595jdaqb6w9og#L3028)?

vtjnash commented:
It seems like something is failing to drop a gc-reference (https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.13610/job/iex595jdaqb6w9og#L2826)

vtjnash committed:
update inInference flag usage, enable infering inference

it now just needs to be a hint to jl_type_infer not to bother trying to infer a method
(helps avoid accidental recursion over inferring type inference)

vtjnash committed:
inference: de-thread parent parameter

it's a bit cleaner to handle this in the parent
(now that recursion detection is so thoroughly decoupled)

also restores parent backedge lost from typeinf_edge

vtjnash committed:
ensure cycle-detection looks at the whole cycle set

martinholters commented on base/inference.jl:
```diff
@@ -1400,10 +1398,6 @@ function abstract_call_gf_by_type(f::ANY, atype::ANY, sv::InferenceState)
     end
-    if isempty(applicable)
-        # TODO: this is needed because type intersection is wrong in some cases
-        return Any
```
When I proposed this in #20866 you seemed rather skeptical. Now it's my turn to be skeptical :smile:
There's no doubt we want this in the long run, but I now think it's still too early because type intersection is still too buggy. (We don't need the extra crashes produced by this to raise extra awareness of these bugs, they are still easy enough to find anyway.)

And I wouldn't want this PR to be delayed due to open type intersection bugs.

For the record: This is responsible for the crash in the arrayops test.

martinholters replied:
Oh, it's not type intersection this time (the failing arrayops test), see #21769. Indeed, there should be a `MethodError`, but there isn't.

martinholters replied:
With
```diff
diff --git a/test/arrayops.jl b/test/arrayops.jl
index c6851c018f..5b4db92fdc 100644
--- a/test/arrayops.jl
+++ b/test/arrayops.jl
@@ -2059,12 +2059,12 @@ struct F21666{T <: Base.TypeArithmetic}
     x::Float32
 end
 
+Base.TypeArithmetic(::Type{F21666{T}}) where {T} = T()
+Base.:+(x::F, y::F) where {F <: F21666} = F(x.x + y.x)
+Base.convert(::Type{Float64}, x::F21666) = Float64(x.x)
 @testset "Exactness of cumsum # 21666" begin
     # test that cumsum uses more stable algorithm
     # for types with unknown/rounding arithmetic
-    Base.TypeArithmetic(::Type{F21666{T}}) where {T} = T
-    Base.:+(x::F, y::F) where {F <: F21666} = F(x.x + y.x)
-    Base.convert(::Type{Float64}, x::F21666) = Float64(x.x)
     # we make v pretty large, because stable algorithm may have a large base case
     v = zeros(300); v[1] = 2; v[200:end] = eps(Float32)
 
diff --git a/test/strings/basic.jl b/test/strings/basic.jl
index b45b6757d8..e88f720c7a 100644
--- a/test/strings/basic.jl
+++ b/test/strings/basic.jl
@@ -193,7 +193,7 @@ gstr = GenericString("12")
 @test ind2chr(gstr,2)==2
 
 # issue #10307
-@test typeof(map(Int16,AbstractString[])) == Vector{Int16}
+@test typeof(map(x -> parse(Int16, x), AbstractString[])) == Vector{Int16}
 
 for T in [Int8, UInt8, Int16, UInt16, Int32, UInt32, Int64, UInt64, Int128, UInt128]
     for i in [typemax(T), typemin(T)]
```
this passes locally for me.

vtjnash commented:
Alright, this should be ready to merge. I'll do so over the weekend if there's no concerns.

andyferris commented:
Would it make any sense to backport this to v0.6? 

(Obviously any improvements to inference will improve the v0.6 experience, and I'm guessing *not* backporting this would block backporting any other future improvements)

tkelman commented:
is this bisect clean or should it be squashed?

vtjnash commented:
I think I would recommend squashing it to one commit. I think this can be back-ported remarkably cleanly. It'll probably also handle the backedge computation better (which may fix https://github.com/JuliaLang/julia/issues/21653)

JeffBezanson commented on base/inference.jl:
```diff
@@ -296,10 +305,7 @@ end
 
 #### current global inference state ####
```
Seriously though, would be nice to just delete this section :)

JeffBezanson commented on base/inference.jl:
```diff
@@ -2417,91 +2397,113 @@ function code_for_method(method::Method, atypes::ANY, sparams::SimpleVector, wor
+    end
+end
+
+function resolve_call_cycle!(linfo::MethodInstance, parent::InferenceState)
```
This function could use a comment. Not obvious to me what `resolve` means in this context.

tkelman commented on base/inference.jl:
```diff
@@ -2417,91 +2397,113 @@ function code_for_method(method::Method, atypes::ANY, sparams::SimpleVector, wor
-                return nothing
+function merge_call_chain!(parent::InferenceState, ancestor::InferenceState, child::InferenceState)
+    # add backedge of parent <- child
+    # then add all backeges of parent <- parent.parent
```
backedges

tkelman commented on base/inference.jl:
```diff
@@ -266,27 +265,37 @@ mutable struct InferenceState
-        src = ccall(:jl_uncompress_ast, Any, (Any, Any), li.def, li.def.source)
+function InferenceState(linfo::MethodInstance,
+                        optimize::Bool, cached::Bool, params::InferenceParams)
+    # prepare an InferenceState object for infering lambda
```
inferring

tkelman commented on base/sysimg.jl:
```diff
@@ -73,6 +72,9 @@ include("refpointer.jl")
 include("checked.jl")
 importall .Checked
 
+# buggy handling of ispure in type-inference means this should be after re-defining the basic operations that they might try to call
```
line wrap comment

tkelman commented:
We could still use devdocs explaining what a backedge is

jrevels committed:
remove global state section

jrevels committed:
add explanation of resolve_call_cycle function

jrevels committed:
fix typos and comment length

jrevels commented:
Made the requested clean-ups and added a comment describing `resolve_call_cycle` (@vtjnash might want to check that my explanation is correct/complete).

> We could still use devdocs explaining what a backedge is

AFAIK, a backedge is a reference from one frame to another (within a call graph) that facilitates the reverse-propagation of type information. I'll leave the actual documentation to Jameson, though I imagine that could be done in a separate PR.

vtjnash commented:
> We could still use devdocs explaining what a backedge is

There's several different types. An `edge` is simply a function call. A `backedge` is that same thing, backwards. At some point, I'll get Jeff to sit down for a couple days with me (and anyone else who's interested) and try to figure out what they are.

vtjnash commented:
https://juliacomputing.com/blog/2017/05/15/inference-converage2.html

andyferris commented:
This fixes an intermittent (!!) inference problem that I get on v0.6.0-rc3 on a project I am working on.

Any chance of backporting this to v0.6?

andyferris commented:
Will this one still make it into v0.6-final? (it fixes some problems I'm having with intermittent (i.e. unit tests failing `@inferred` while `@code_warntype` works OK at REPL) inference on rc3).

tkelman commented:
Can you contribute base tests that were fixed by this?
